modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+            otrl_context_set_trust(fprint, "");
+            PutModuleContext(ctx, CString("Fingerprint ") +
+                                      HumanFingerprint(fprint) +
+                                      " distrusted!");
+            WriteFingerprints();
+        }
+    }
+
+    void CmdFinish(const CString& sLine) {
+        ConnContext* ctx = GetContextFromArg(sLine);
+        if (!ctx) {
+            return;
+        }
+
+        otrl_message_disconnect(m_pUserState, &m_xOtrOps, this,
+                                ctx->accountname, PROTOCOL_ID, ctx->username,
+                                ctx->their_instance);
+        PutModuleContext(ctx, "Conversation finished.");
+    }
+
+    void DoSMP(ConnContext* ctx, const CString& sQuestion,
+               const CString& sSecret) {
+        if (sSecret.Equals("")) {
+            PutModuleContext(ctx, "No secret given!");
+            return;
+        }
+
+        if (ctx->msgstate != OTRL_MSGSTATE_ENCRYPTED) {
+            PutModuleContext(ctx, "Not in OTR session.");
+            return;
+        }
+
+        COtrAppData* ad = static_cast<COtrAppData*>(ctx->app_data);
+        assert(ad);
+
+        if (ad->bSmpReply) {
+            otrl_message_respond_smp(
+                m_pUserState, &m_xOtrOps, this, ctx,
+                reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                sSecret.length());
+            PutModuleContext(ctx, "Responded to authentication.");
+        } else {
+            if (sQuestion.Equals("")) {
+                otrl_message_initiate_smp(
+                    m_pUserState, &m_xOtrOps, this, ctx,
+                    reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                    sSecret.length());
+            } else {
+                otrl_message_initiate_smp_q(
+                    m_pUserState, &m_xOtrOps, this, ctx, sQuestion.c_str(),
+                    reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                    sSecret.length());
+            }
+            PutModuleContext(ctx, "Initiated authentication.");
+        }
+
+        ad->bSmpReply = false;
+    }
+
+    void CmdAuth(const CString& sLine) {
+        ConnContext* ctx = GetContextFromArg(sLine);
+        if (!ctx) {
+            return;
+        }
+
+        CString sSecret = sLine.Token(2, true);
+        DoSMP(ctx, "", sSecret);
+    }
+
+    void CmdAuthQ(const CString& sLine) {
+        ConnContext* ctx = GetContextFromArg(sLine);
+        if (!ctx) {
+            return;
+        }
+
+        COtrAppData* ad = static_cast<COtrAppData*>(ctx->app_data);
+        assert(ad);
+        if (ad->bSmpReply) {
+            PutModuleContext(ctx,
+                             "Authentication in progress. Use Auth to "
+                             "respond with secret, or AuthAbort to abort it");
+            return;
+        }
+
+        CString sRest = sLine.Token(2, true);
+        if (sRest.length() == 0 || sRest[0] != '[') {
```

Comment:
The AuthQ syntax is `'authq' NICK '[' QUESTION ']' ANSWER` where both QUESTION and ANSWER can be composed of multiple words. The only reason [] are used is because they are used in the irssi otr plugin.